pepper-apis/src/main/java/org/broadinstitute/ddp/filter/StudyLanguageResolutionFilter.java

```diff
@@ -0,0 +1,120 @@
+
+/**
+ * Figures out a preferred user language taking into account the language weights
+ * from Accept-Language header, information from the user profile and languages
+ * supported by the study. Puts that language into the attribute storage to make
+ * it available later in all routes interested in fetching translated entities
+ */
+public class StudyLanguageResolutionFilter implements Filter {
+
+    private static Map<Locale, LanguageDto> localeToLanguage = new HashMap<>();
+
+    public static final String USER_LANGUAGE = "USER_LANGUAGE";
+    private static final Logger LOG = LoggerFactory.getLogger(StudyLanguageResolutionFilter.class);
+    private static final String STUDY_GUID_REGEX = "/studies/(\\w+)";
+    private static final int STUDY_GUID_INDEX = 1;
+
+    @Override
+    public void handle(Request request, Response response) {
+        try {
+            String acceptLanguageHeader = request.headers(RouteConstants.ACCEPT_LANGUAGE);
+            Matcher matcher = Pattern.compile(STUDY_GUID_REGEX).matcher(request.url());
+            String studyGuid = matcher.find() ? matcher.group(STUDY_GUID_INDEX) : null;
+            // The "supported languages" notion is an attribute of a study, thus is doesn't
+            // make any sense outside of the study context, so we issue a warning
+            boolean supportedLanguagesCanBeDetected = studyGuid != null;
+            if (!supportedLanguagesCanBeDetected) {
+                LOG.warn(
+                        "Supported languages can't be detected because the filter is invoked"
+                        + " before the route that is outside of the study context. Please"
+                        + " remount the filter under '/user/*/studies/*' instead. Current path = {}",
+                        request.url()
+                );
+                return;
+            }
+            LOG.info("The supported languages can be detected for the study {}", studyGuid);
+            Locale ddpAuthPreferredLocale = RouteUtil.getDDPAuth(request).getPreferredLocale();
+            TransactionWrapper.useTxn(
+                    handle -> {
+                        Locale preferredLocale = StudyLanguageResolutionFilter.getPreferredLocale(
+                                handle, acceptLanguageHeader, ddpAuthPreferredLocale, studyGuid
+                        );
+                        LanguageDto preferredLanguage = StudyLanguageResolutionFilter.convertLocaleToLanguageDto(
+                                handle, preferredLocale
+                        );
+                        request.attribute(USER_LANGUAGE, preferredLanguage);
+                        LOG.info("Added the preferred user language {} to the attribute store", preferredLanguage.getIsoCode());
+                        String contentLanguageHeader = StudyLanguageResolutionFilter.createContentLanguageHeaderFromLocale(preferredLocale);
+                        response.header(RouteConstants.CONTENT_LANGUAGE, contentLanguageHeader);
```

Comment:
Will this mean that we'll always add the `content-language` header? What happens if the route itself throws an error or returns an error response? Does it make sense to include `content-langauage` header in that case?

If not, maybe one idea is to have an `after()` filter to handle this? I.e. `StudyContentLanguageFilter` will just need to look if `response.statusCode()` is `2xx` and `request.attribute()` has `user_language`, then we tag on a `content-language`. Or maybe leave it up to the individual routes to add this header?

Reply:
I chose the last option: checking for the status code == 200 and the presence of `USER_LANGUAGE` attribute.

Reply:
Hmm, saw latest change but not sure how it solves issue... This filter runs _before_ the actual route handlers, so this means we have set the `content-language` header even before the route code executes. Did you mean to pull this code out to a separate filter class that's mounted in a `after()`?

Reply:
Yes, you're right, I overlooked the fact that unlike classic servlet filters, Spark `before` filters are executed only **before** the route. Split the filter into two and mounted the new filter in an `after()`.